pkg/kv/kvserver/replica_init.go

```diff
@@ -395,6 +395,15 @@ func (r *Replica) setDescLockedRaftMuLocked(ctx context.Context, desc *roachpb.R
 	r.concMgr.OnRangeDescUpdated(desc)
 	r.mu.state.Desc = desc
 
+	// Determine whether the replica unconditionally requires an expiration-based
+	// lease. Ranges located before or including the node liveness table must
+	// always use expiration leases to avoid circular dependencies on the node
+	// liveness table.
+	//
+	// Some tests don't use node liveness, and require expiration leases.
+	r.mu.requiresExpirationLease = r.store.cfg.NodeLiveness == nil ||
```

Comment:
I audited the code, `setDescLockedRaftMuLocked` is always invoked when we change the replica's `Desc`, so this seems indeed the only place where we should update `requiresExpirationLease`.

Technically, the assigning of `Desc` happens in other places, but it's alway coupled with `setDescLockedRaftMuLocked` in the same block.

Reply:
I mean, there are a few places which assign the whole `r.mu.state` (which includes `Desc`), but only 2 cases are interesting: where `Desc` is initialized (which also means it has a key span). These are split and snapshot apply paths. In both of them loading `r.mu.state` is coupled with `setDescLockedRaftMuLocked `.